### DISCUSSION TITLE
OnInitialActivate does not get called in "Child" Screen

### DISCUSSION THREAD

**Discussion by uyaem** [canton7/Stylet#276] [Q&A] [answered]
**Description**
OnInitialActivate does not get called in a "Child" Screen in controls if using `<ContentControl>` to embed the child.
I cannot use any of the Conductor<T> options on the parent, because there are potentially several "dynamic controls" on the page, as the application continues to grow.
Is this a bug, or by design, or am I doing something wrong?

**To Reproduce**
Below code snippet should suffice to reproduce the behaviour.
This is an extremely stripped down version of my scenario, and I hope it is complete enough. If not, I'll amend as needed.

```xml
/// ParentView.xaml
[...]
    <ContentControl s:View.Model="{Binding DynamicControl}" [...] />
[...]
```

```csharp
/// ParentViewModel.cs
public class ParentViewModel : Screen
{
  private Func<ChildViewModel> createChild;
  private Screen dynamicControl;

  public ParentViewModel(Func<ChildViewModel> createChild) { this.createChild = createChild; }

  public Screen DynamicControl
  {
    get => this.dynamicControl;
    set => this.SetAndNotify(ref this.dynamicControl, value);
  }

  protected override void OnInitialActivate()
  {
    base.OnInitialActivate();
    this.DynamicControl = this.createChild();
  }
}

/// ChildViewModel.cs
public class ChildViewModel : Screen
{
  protected override void OnInitialActivate()
  {
    // Never gets called
  }
}
```

**Version Info**
 - Stylet version: 1.3.6
 - Runtime version: .NET 5

**Comment by canton7** [comment, marked as answer]
(Converted this support request to a discussion. Please read the issue template).

The parent doesn't know that the child exists. How would it? It's therefore impossible to know that there's something which needs to be activated.

You can use a `Conductor<IScreen>`, and set `ActiveItem` rather than having a `DynamicItem` property: the conductor knows that the `ActiveItem` property contains a child which needs to be activated/etc. [See the documentation](https://github.com/canton7/Stylet/wiki/Screens-and-Conductors).

You can also use `child.ConductWith(parent)` for long-lived children, but don't do this if the child might be destroyed before the parent.

**Reply by uyaem** [reply]
Thanks for your fast reply.

Because there will be multiple dynamic controls, `Conductor<T>` was not an option - but `Conductor<T>.Collection.AllActive` is! (and already works like a charm...)

For the sake of completion, and if others are facing the same task, here's the altered code:
```csharp
/// ParentViewModel.cs
public class ParentViewModel : Conductor<IScreen>.Collection.AllActive
{
  [...]

  protected override void OnInitialActivate()
  {
    base.OnInitialActivate();
    this.Items.Add(this.createChildOne());
    this.Items.Add(this.createChildTwo());
  }
}
```

```xml
/// ParentView.xaml
[...]
  <ContentControl s:View.Model="{Binding Items[0]}" [...] />
  <ContentControl s:View.Model="{Binding Items[1]}" [...] />
[...]
```

**Reply by canton7** [reply]
That's why writing a representative sample is useful :)

If your children actually live as long as the parent, then I'd write:

```cs
public class ParentViewModel : Screen
{
    public ChildOneViewModel ChildOne { get; }
    public ChildTwoViewModel ChildTwo { get; }

    public ParentViewModel(ChildOneViewModel childOne, ChildTwoViewModel childTwo)
    {
        ChildOne = childOne;
        ChildOne.ConductWith(this);

        ChildTwo = childTwo;
        ChildTwo.ConductWith(this);
    }
}
```

That gives you named properties to bind to, rather than the opaque `Items[0]`, etc. The downside is that it's hard to undo the `ConductWith` if you need to destroy the child: it uses weak events, but there will still be a period where the child may be activated/etc, before the GC gets it.

**Reply by uyaem** [reply]
I actually did this - I don't trust myself messing up the indices eventually ;)
Just still using constructor injection, as my screens have injected dependencies themselves, so the Func<T> factories come in very handy.

**Reply by canton7** [reply]
(Injecting a `ChildOneViewModel` should work as well as injecting a `Func<ChildOneViewModel>`, if you don't need to create multiple instances / don't *need* the instantiation of `ChildOneViewModel` to be delayed for some reason)
